README.md
# test_coverage [![Build Status](https://travis-ci.com/pulyaevskiy/test-coverage.svg?branch=master)](https://travis-ci.com/pulyaevskiy/test-coverage) [![Pub](https://img.shields.io/pub/v/test_coverage.svg?style=flat)](https://pub.dartlang.org/packages/test_coverage) ![Coverage](https://raw.githubusercontent.com/pulyaevskiy/test-coverage/master/coverage_badge.svg?sanitize=true)

A simple command-line tool to collect test coverage information from
Dart VM tests. It is useful if you need to generate coverage reports
locally during development.

## Usage

Add dev dependency to your `pubspec.yaml`:

```yaml
dev_dependencies:
  test_coverage:
    git: https://github.com/priezz/test-coverage
```

Run `pub get` to install. Then, in the root of your project run:

```bash
pub run test_coverage
```

Result is saved in `coverage/lcov.info`. If you have `lcov` tool
installed on your system (for Mac it's `brew install lcov`) you can
generate coverage reports using `genhtml` command:

```bash
genhtml -o coverage coverage/lcov.info
# Open in the default browser (mac):
open coverage/index.html
```

## Integrations

Resulting `coverage/lcov.info` file is ready to be consumed by
Codecov command-line tool, so no extra step is needed.

This library was not tested with coveralls yet.

## Generating badge image

![Coverage](https://raw.githubusercontent.com/pulyaevskiy/test-coverage/master/coverage_badge.svg?sanitize=true)

Coverage badge SVG image is automatically generated and saved to `coverage_badge.svg` in your
project root directory. You can add it to Git and use in README.md on Github as follows:

```md
![Coverage](https://raw.githubusercontent.com/{you}/{repo}/master/coverage_badge.svg?sanitize=true)
```

If you don't need it you can pass `--no-badge` flag when running test coverage.

## Known limitations

* This library was created to run Dart VM tests. It has not been tested
  and likely won't work for Dart code targeting web platform (compiled
  to JavaScript). There is no need to use this tool for Flutter as it
  allows collecting coverage information with `flutter test --coverage`.

## How it works

The tool performs following steps:

### 1. Generates `test/.test_coverage.dart` file (a "test all" script).

Scans your `test/` directory to find all `*_test.dart` files and creates `test/.test_coverage.dart`
which imports all found test files and runs corresponding `main()` of all tests functions within
the same file (and process), which simplifies collection of coverage information.

It is recommended to add this file to your `.gitignore`.

Below is an example of `test/.test_coverage.dart`:

```dart
// Auto-generated by test_coverage. Do not edit by hand.
// Consider adding this file to your .gitignore.

import 'some_test.dart' as some_test;
import 'nested/other_test.dart' as other_test;
import 'some_other_test.dart' as some_other_test;

void main() {
  some_test.main();
  other_test.main();
  some_other_test.main();
}
```

### 2. Runs the tests

Following command is used to run the tests:

```
dart --pause-isolates-on-exit --enable_asserts --enable-vm-service \
  test/.test_coverage.dart
```

### 3. Collects and formats coverage information

When test execution is completed the tool uses functionality of the
`coverage` package to collect and format coverage report.

Feel free to file feature requests and bug reports at the
[issue tracker][].

[issue tracker]: https://github.com/pulyaevskiy/test-coverage/issues

### 4. Minimal coverage percentage

Set minimal coverage percentage to pass
`min-coverage` percentage value of coverage.

### 5. See test output
`print-test-output` to show test output.

Show output of tests

Feel free to file feature requests and bug reports at the
[issue tracker][].

[issue tracker]: https://github.com/pulyaevskiy/test-coverage/issues


CHANGELOG.md
# CHANGELOG

## 0.8.0

- Write pretty output to `coverage.txt`

## 0.7.0

- Upgrade dependencies
- Refactor to stop using deprecated coverage methods

## 0.6.0

- Migrated code to null safety (#33)
- Removed discontinued lcov dependency (#38)

## 0.5.0

- Updated dependency on coverage package to ^0.14.1 (#24)
- Badge generation now honours the corresponding command line argument (#27)

## 0.4.3

- Fix support for --[no-]badge. (#21)

## 0.4.2

* Minimal coverage percentage & test output to console

## 0.4.1

* Fixed test generator for windows platform (#11)

## 0.4.0

* Upgraded to coverage 0.13.0 (#8)

## 0.3.0

* Refactor code for Dart 2.0 features.
* Added `--exclude`, `--port`, `--help` and `--[no-]badge` options. See `pub run test_coverage -h`
  for more details.
* Coverage badge moved from `coverage/badge.svg` to `coverage_badge.svg` (in the package root)
  which makes it easier to manage `coverage/` in `.gitignore`, allows you to commit the badge
  to your repo and include it in the `README.md`.

## 0.2.4

* Upgraded dependencies.
* Bumped SDK constraint to `2.3.0`.

## 0.2.3

* Prepare for Dart 2 stable.

## 0.2.2

* Renamed `coverage/coverage.lcov` to `coverage/lcov.info`.

## 0.2.1

* Generate coverage badge in `coverage/badge.svg`.

## 0.2.0

* Moved generated reports to `coverage/` subfolder.
* Added tests.
* Updated usage instructions to recommend local `dev_dependencies`
  and `pub run test_coverage` instead of global installation.

## 0.1.1

* Make paths in lcov file relative to the project root.

## 0.1.0

* Initial version


.history/lib/src/functions_20210412121035.dart
// Copyright (c) 2018, Anatoly Pulyaevskiy. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:coverage/coverage.dart' as coverage;
import 'package:glob/glob.dart';
import 'package:lcov_dart/lcov_dart.dart';
import 'package:path/path.dart' as path;

final _sep = path.separator;

List<File> findTestFiles(Directory packageRoot, {Glob excludeGlob}) {
  final testsPath = path.join(packageRoot.absolute.path, 'test');
  final testsRoot = Directory(testsPath);
  final contents = testsRoot.listSync(recursive: true);
  final result = <File>[];
  for (final item in contents) {
    if (item is! File) continue;
    if (!item.path.endsWith('_test.dart')) continue;
    final relativePath = item.path.substring(packageRoot.path.length + 1);
    if (excludeGlob != null && excludeGlob.matches(relativePath)) {
      continue;
    }
    result.add(item);
  }
  return result;
}

class TestFileInfo {
  final File testFile;
  final String alias;
  final String import;

  TestFileInfo._(this.testFile, this.alias, this.import);

  factory TestFileInfo.forFile(File testFile) {
    final parts = testFile.absolute.path.split(_sep).toList();
    var relative = <String>[];
    while (parts.last != 'test') {
      relative.add(parts.last);
      parts.removeLast();
    }
    relative = relative.reversed.toList();
    final alias = relative.join('_').replaceFirst('.dart', '');
    final importPath = relative.join('/');
    final import = "import '$importPath' as $alias;";
    return TestFileInfo._(testFile, alias, import);
  }
}

void generateMainScript(Directory packageRoot, List<File> testFiles) {
  final imports = <String>[];
  final mainBody = <String>[];

  for (final test in testFiles) {
    final info = TestFileInfo.forFile(test);
    imports.add(info.import);
    mainBody.add('  ${info.alias}.main();');
  }
  imports.sort();

  final buffer = StringBuffer()
    ..writeln('// Auto-generated by test_coverage. Do not edit by hand.')
    ..writeln('// Consider adding this file to your .gitignore.')
    ..writeln();
  imports.forEach(buffer.writeln);
  buffer..writeln()..writeln('void main() {');
  mainBody.forEach(buffer.writeln);
  buffer.writeln('}');
  File(
    path.join(packageRoot.path, 'test', '.test_coverage.dart'),
  ).writeAsStringSync(buffer.toString());
}

Future<void> runTestsAndCollect(String packageRoot, String port,
    {bool printOutput = false}) async {
  final script = path.join(packageRoot, 'test', '.test_coverage.dart');
  final dartArgs = [
    '--pause-isolates-on-exit',
    '--enable_asserts',
    '--enable-vm-service=$port',
    script
  ];
  final process =
      await Process.start('dart', dartArgs, workingDirectory: packageRoot);
  final serviceUriCompleter = Completer<Uri>();
  process.stdout
      .transform(utf8.decoder)
      .transform(const LineSplitter())
      .listen((line) {
    if (printOutput) print(line);
    if (serviceUriCompleter.isCompleted) return;
    final uri = _extractObservatoryUri(line);
    if (uri != null) {
      serviceUriCompleter.complete(uri);
    } else {
      serviceUriCompleter.completeError(line);
    }
  });

  final serviceUri = await serviceUriCompleter.future.catchError((error) {
    process.kill(ProcessSignal.sigkill);
  });

  if (serviceUri == null) {
    throw StateError('Could not run tests with Observatory enabled. '
        'Try setting a different port with --port option.');
  }

  Map<String, Map<int, int>> hitmap;
  try {
    final data = await coverage.collect(serviceUri, true, true, false, {});
    hitmap = await coverage.createHitmap(data['coverage']);
  } finally {
    await process.stderr.drain<List<int>>();
  }
  final exitStatus = await process.exitCode;
  if (exitStatus != 0) {
    throw 'Tests failed with exit code $exitStatus';
  }
  final resolver = coverage.Resolver(
    packagesPath: path.join(packageRoot, '.packages'),
  );
  final lcov = coverage.LcovFormatter(
    resolver,
    reportOn: ['lib${path.separator}'],
    basePath: packageRoot,
  );
  final coverageData = await lcov.format(hitmap);
  final coveragePath = path.join(packageRoot, 'coverage');
  final coverageDir = Directory(coveragePath);
  if (!coverageDir.existsSync()) {
    coverageDir.createSync();
  }
  File(path.join(coveragePath, 'lcov.info')).writeAsStringSync(coverageData);
}

// copied from `coverage` package
Uri _extractObservatoryUri(String str) {
  const kObservatoryListening = 'Observatory listening on ';
  final msgPos = str.indexOf(kObservatoryListening);
  if (msgPos == -1) return null;
  final startPos = msgPos + kObservatoryListening.length;
  final endPos = str.indexOf(RegExp(r'(\s|$)'), startPos);
  try {
    return Uri.parse(str.substring(startPos, endPos));
  } on FormatException {
    return null;
  }
}

double calculateLineCoverage(File lcovReport) {
  final report = Report.fromCoverage(lcovReport.readAsStringSync());
  var totalLines = 0;
  var hitLines = 0;
  for (final rec in report.records) {
    for (final line in rec.lines.data) {
      totalLines++;
      hitLines += (line.executionCount > 0) ? 1 : 0;
    }
  }
  return hitLines / totalLines;
}

void generateBadge(Directory packageRoot, double lineCoverage) {
  const leftWidth = 59;
  final value = '${(lineCoverage * 100).floor()}%';
  final color = _color(lineCoverage);
  final metrics = _BadgeMetrics.forPercentage(lineCoverage);
  final rightWidth = metrics.width - leftWidth;
  final content = _kBadgeTemplate
      .replaceAll('{width}', metrics.width.toString())
      .replaceAll('{rightWidth}', rightWidth.toString())
      .replaceAll('{rightX}', metrics.rightX.toString())
      .replaceAll('{rightLength}', metrics.rightLength.toString())
      .replaceAll('{color}', color.toString())
      .replaceAll('{value}', value.toString());
  File(path.join(packageRoot.path, 'coverage_badge.svg'))
      .writeAsStringSync(content);
}

class _BadgeMetrics {
  final int width;
  final int rightX;
  final int rightLength;

  _BadgeMetrics({this.width, this.rightX, this.rightLength});

  factory _BadgeMetrics.forPercentage(double value) {
    final pct = (value * 100).floor();
    if (pct.toString().length == 1) {
      return _BadgeMetrics(
        width: 88,
        rightX: 725,
        rightLength: 190,
      );
    } else if (pct.toString().length == 2) {
      return _BadgeMetrics(
        width: 94,
        rightX: 755,
        rightLength: 250,
      );
    } else {
      return _BadgeMetrics(
        width: 102,
        rightX: 795,
        rightLength: 330,
      );
    }
  }
}

String _color(double percentage) {
  final map = {
    0.0: _Color(0xE0, 0x5D, 0x44),
    0.5: _Color(0xE0, 0x5D, 0x44),
    0.6: _Color(0xDF, 0xB3, 0x17),
    0.9: _Color(0x97, 0xCA, 0x00),
    1.0: _Color(0x44, 0xCC, 0x11),
  };
  double lower;
  double upper;
  for (final key in map.keys) {
    if (percentage < key) {
      upper = key;
      break;
    }
    if (key < 1.0) lower = key;
  }
  upper ??= 1.0;
  final lowerColor = map[lower];
  final upperColor = map[upper];
  final range = upper - lower;
  final rangePct = (percentage - lower) / range;
  final pctLower = 1 - rangePct;
  final pctUpper = rangePct;
  final r = (lowerColor.r * pctLower + upperColor.r * pctUpper).floor();
  final g = (lowerColor.g * pctLower + upperColor.g * pctUpper).floor();
  final b = (lowerColor.b * pctLower + upperColor.b * pctUpper).floor();
  final color = _Color(r, g, b);
  return color.toString();
}

class _Color {
  final int r, g, b;

  _Color(this.r, this.g, this.b);

  @override
  String toString() =>
      '#${((1 << 24) + (r << 16) + (g << 8) + b).toRadixString(16).substring(1)}';
}

const _kBadgeTemplate = '''
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="{width}" height="20">
  <linearGradient id="b" x2="0" y2="100%">
    <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
    <stop offset="1" stop-opacity=".1"/>
  </linearGradient>
  <clipPath id="a">
    <rect width="{width}" height="20" rx="3" fill="#fff"/>
  </clipPath>
  <g clip-path="url(#a)">
    <path fill="#555" d="M0 0h59v20H0z"/>
    <path fill="{color}" d="M59 0h{rightWidth}v20H59z"/>
    <path fill="url(#b)" d="M0 0h{width}v20H0z"/>
  </g>
  <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110">
    <text x="305" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="490">coverage</text>
    <text x="305" y="140" transform="scale(.1)" textLength="490">coverage</text>
    <text x="{rightX}" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="{rightLength}">{value}</text>
    <text x="{rightX}" y="140" transform="scale(.1)" textLength="{rightLength}">{value}</text>
  </g>
</svg>
''';


lib/src/functions.dart
// Copyright (c) 2018, Anatoly Pulyaevskiy. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:coverage/coverage.dart' as coverage;
import 'package:glob/glob.dart';
import 'package:path/path.dart' as path;

final _sep = path.separator;

List<File> findTestFiles(Directory packageRoot, {Glob? excludeGlob}) {
  final testsPath = path.join(packageRoot.absolute.path, 'test');
  final testsRoot = Directory(testsPath);
  final contents = testsRoot.listSync(recursive: true);
  final result = <File>[];
  for (final item in contents) {
    if (item is! File) continue;
    if (!item.path.endsWith('_test.dart')) continue;
    final relativePath = item.path.substring(packageRoot.path.length + 1);
    if (excludeGlob != null && excludeGlob.matches(relativePath)) {
      continue;
    }
    result.add(item);
  }
  return result;
}

class TestFileInfo {
  final File testFile;
  final String alias;
  final String import;

  TestFileInfo._(this.testFile, this.alias, this.import);

  factory TestFileInfo.forFile(File testFile) {
    final parts = testFile.absolute.path.split(_sep).toList();
    var relative = <String>[];
    while (parts.last != 'test') {
      relative.add(parts.last);
      parts.removeLast();
    }
    relative = relative.reversed.toList();
    final alias = relative.join('_').replaceFirst('.dart', '');

    final importPath = relative.join(path.separator);
    final import = "import '$importPath' as $alias;";
    return TestFileInfo._(testFile, alias, import);
  }
}

void generateMainScript(Directory packageRoot, List<File> testFiles) {
  final imports = <String>[];
  final mainBody = <String>[];

  for (final test in testFiles) {
    final info = TestFileInfo.forFile(test);
    imports.add(info.import);
    mainBody.add('  ${info.alias}.main();');
  }
  imports.sort();

  final buffer = StringBuffer()
    ..writeln('// Auto-generated by test_coverage. Do not edit by hand.')
    ..writeln('// Consider adding this file to your .gitignore.')
    ..writeln();
  imports.forEach(buffer.writeln);
  buffer
    ..writeln()
    ..writeln('void main() {');
  mainBody.forEach(buffer.writeln);
  buffer.writeln('}');
  File(
    path.join(packageRoot.path, 'test', '.test_coverage.dart'),
  ).writeAsStringSync(buffer.toString());
}

Future<void> runTestsAndCollect(String packageRoot, String? port,
    {bool printOutput = false}) async {
  final script = path.join(packageRoot, 'test', '.test_coverage.dart');
  final dartArgs = [
    // '--pause-isolates-on-exit',
    '--enable_asserts',
    '--enable-vm-service=$port',
    script
  ];
  final process =
      await Process.start('dart', dartArgs, workingDirectory: packageRoot);
  final serviceUriCompleter = Completer<Uri?>();
  process.stdout
      .transform(utf8.decoder)
      .transform(const LineSplitter())
      .listen((line) {
    if (printOutput) print(line);
    if (serviceUriCompleter.isCompleted) return;
    final uri = _extractObservatoryUri(line);
    if (uri != null) {
      serviceUriCompleter.complete(uri);
    } else {
      serviceUriCompleter.completeError(line);
    }
  });

  final serviceUri = await serviceUriCompleter.future.catchError((error) {
    process.kill(ProcessSignal.sigkill);
  });

  if (serviceUri == null) {
    throw StateError('Could not run tests with Observatory enabled. '
        'Try setting a different port with --port option.');
  }

  Map<String, coverage.HitMap> hitmap;
  try {
    final Map<String, dynamic> data =
        await coverage.collect(serviceUri, true, true, false, {});
    hitmap = await coverage.HitMap.parseJson(data['coverage']);
  } finally {
    await process.stderr.drain<List<int>?>();
  }
  final int exitStatus = await process.exitCode;
  if (exitStatus != 0) throw 'Tests failed with exit code $exitStatus';

  final coverage.Resolver resolver = await coverage.Resolver.create(
    packagesPath: path.join(packageRoot, '.packages'),
  );
  final String coverageData = hitmap.formatLcov(
    resolver,
    basePath: packageRoot,
    reportOn: ['lib${path.separator}'],
  );
  final String coveragePath = path.join(packageRoot, 'coverage');
  final Directory coverageDir = Directory(coveragePath);
  if (!coverageDir.existsSync()) coverageDir.createSync();
  File(path.join(coveragePath, 'lcov.info')).writeAsStringSync(coverageData);

  final String prettyData = await hitmap.prettyPrint(
    resolver,
    coverage.Loader(),
    reportBranches: true,
    reportFuncs: true,
    reportOn: ['lib${path.separator}'],
  );
  File(path.join(coveragePath, 'coverage.txt')).writeAsStringSync(prettyData);
}

// copied from `coverage` package
Uri? _extractObservatoryUri(String str) {
  const kObservatoryListening = 'Observatory listening on ';
  final msgPos = str.indexOf(kObservatoryListening);
  if (msgPos == -1) return null;
  final startPos = msgPos + kObservatoryListening.length;
  final endPos = str.indexOf(RegExp(r'(\s|$)'), startPos);
  try {
    return Uri.parse(str.substring(startPos, endPos));
  } on FormatException {
    return null;
  }
}

double calculateLineCoverage(File lcovReport) {
  final lflhRegex = RegExp(r'^L[FH]:(\d*)$');
  final reportFileContent = lcovReport.readAsLinesSync();
  reportFileContent.retainWhere((l) => lflhRegex.hasMatch(l));
  var totalLines = 0;
  var hitLines = 0;
  for (final line in reportFileContent) {
    var valueString = lflhRegex.matchAsPrefix(line)?.group(1) ?? '0';
    var value = int.tryParse(valueString) ?? 0;
    if (line.startsWith('LF')) {
      totalLines += value;
    } else {
      hitLines += value;
    }
  }
  if (totalLines == 0) return 1;
  return hitLines / totalLines;
}

void generateBadge(Directory packageRoot, double lineCoverage) {
  const leftWidth = 59;
  final value = '${(lineCoverage * 100).floor()}%';
  final color = _color(lineCoverage);
  final metrics = _BadgeMetrics.forPercentage(lineCoverage);
  final rightWidth = metrics.width - leftWidth;
  final content = _kBadgeTemplate
      .replaceAll('{width}', metrics.width.toString())
      .replaceAll('{rightWidth}', rightWidth.toString())
      .replaceAll('{rightX}', metrics.rightX.toString())
      .replaceAll('{rightLength}', metrics.rightLength.toString())
      .replaceAll('{color}', color.toString())
      .replaceAll('{value}', value.toString());
  File(path.join(packageRoot.path, 'coverage_badge.svg'))
      .writeAsStringSync(content);
}

class _BadgeMetrics {
  final int width;
  final int rightX;
  final int rightLength;

  _BadgeMetrics({
    required this.width,
    required this.rightX,
    required this.rightLength,
  });

  factory _BadgeMetrics.forPercentage(double value) {
    final pct = (value * 100).floor();
    if (pct.toString().length == 1) {
      return _BadgeMetrics(
        width: 88,
        rightX: 725,
        rightLength: 190,
      );
    } else if (pct.toString().length == 2) {
      return _BadgeMetrics(
        width: 94,
        rightX: 755,
        rightLength: 250,
      );
    } else {
      return _BadgeMetrics(
        width: 102,
        rightX: 795,
        rightLength: 330,
      );
    }
  }
}

String _color(double percentage) {
  final map = {
    0.0: _Color(0xE0, 0x5D, 0x44),
    0.5: _Color(0xE0, 0x5D, 0x44),
    0.6: _Color(0xDF, 0xB3, 0x17),
    0.9: _Color(0x97, 0xCA, 0x00),
    1.0: _Color(0x44, 0xCC, 0x11),
  };
  late double lower;
  double? upper;
  for (final key in map.keys) {
    if (percentage < key) {
      upper = key;
      break;
    }
    if (key < 1.0) lower = key;
  }
  upper ??= 1.0;
  final lowerColor = map[lower]!;
  final upperColor = map[upper]!;
  final range = upper - lower;
  final rangePct = (percentage - lower) / range;
  final pctLower = 1 - rangePct;
  final pctUpper = rangePct;
  final r = (lowerColor.r * pctLower + upperColor.r * pctUpper).floor();
  final g = (lowerColor.g * pctLower + upperColor.g * pctUpper).floor();
  final b = (lowerColor.b * pctLower + upperColor.b * pctUpper).floor();
  final color = _Color(r, g, b);
  return color.toString();
}

class _Color {
  final int r, g, b;
  _Color(this.r, this.g, this.b);

  @override
  String toString() =>
      '#${((1 << 24) + (r << 16) + (g << 8) + b).toRadixString(16).substring(1)}';
}

const _kBadgeTemplate = '''
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="{width}" height="20">
  <linearGradient id="b" x2="0" y2="100%">
    <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
    <stop offset="1" stop-opacity=".1"/>
  </linearGradient>
  <clipPath id="a">
    <rect width="{width}" height="20" rx="3" fill="#fff"/>
  </clipPath>
  <g clip-path="url(#a)">
    <path fill="#555" d="M0 0h59v20H0z"/>
    <path fill="{color}" d="M59 0h{rightWidth}v20H59z"/>
    <path fill="url(#b)" d="M0 0h{width}v20H0z"/>
  </g>
  <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110">
    <text x="305" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="490">coverage</text>
    <text x="305" y="140" transform="scale(.1)" textLength="490">coverage</text>
    <text x="{rightX}" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="{rightLength}">{value}</text>
    <text x="{rightX}" y="140" transform="scale(.1)" textLength="{rightLength}">{value}</text>
  </g>
</svg>
''';
